sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/OptimizeCsvJsonExprs.scala

```diff
@@ -0,0 +1,116 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.catalyst.optimizer
+
+import org.apache.spark.sql.catalyst.expressions._
+import org.apache.spark.sql.catalyst.plans.logical.LogicalPlan
+import org.apache.spark.sql.catalyst.rules.Rule
+import org.apache.spark.sql.internal.SQLConf
+import org.apache.spark.sql.types.{ArrayType, StructType}
+
+/**
+ * Simplify redundant csv/json related expressions.
+ *
+ * The optimization includes:
+ * 1. JsonToStructs(StructsToJson(child)) => child.
+ * 2. Prune unnecessary columns from GetStructField/GetArrayStructFields + JsonToStructs.
+ * 3. CreateNamedStruct(JsonToStructs(json).col1, JsonToStructs(json).col2, ...) =>
+ *      If(IsNull(json), nullStruct, KnownNotNull(JsonToStructs(prunedSchema, ..., json)))
+ *      if JsonToStructs(json) is shared among all fields of CreateNamedStruct. `prunedSchema`
+ *      contains all accessed fields in original CreateNamedStruct.
+ * 4. Prune unnecessary columns from GetStructField + CsvToStructs.
+ */
+object OptimizeCsvJsonExprs extends Rule[LogicalPlan] {
+  override def apply(plan: LogicalPlan): LogicalPlan = plan transform {
+    case p =>
+      val optimized = if (SQLConf.get.jsonExpressionOptimization) {
+        p.transformExpressions(jsonOptimization)
+      } else {
+        p
+      }
+
+      if (SQLConf.get.csvExpressionOptimization) {
+        optimized.transformExpressions(csvOptimization)
+      } else {
+        optimized
+      }
+  }
```

Comment:
What about:

```scala
case p if SQLConf.get.jsonExpressionOptimization => p.transformExpressions(jsonOptimization)
case p if SQLConf.get.csvExpressionOptimization => p.transformExpressions(csvOptimization)
case p => p
```
? Or is the optimization dependent on each other?

Reply:
Not dependent, but if so `p` only goes to match first case when both are enabled.

Reply:
Ah I see. Since we generalized the rule, it might more make sense to have one configuration to enable and disable this ..

Reply:
To enable/disable entire rule, `spark.sql.optimizer.excludedRules` should be able to help. So the purpose here is to have individual configs for enabling csv/json.

Reply:
Hm .. okay. I still prefer to separate refactoring from this PR but it's fine if both you or @dongjoon-hyun prefer.

Reply:
Technically, I'm neutral in terms of separating.